File: src/store.zig
//! Atomic episode publication under a contained journal root.
//!
//! Default layout: YYYY/MM/DD/<episode-id>.md. Non-default classifications
//! add reserved worlds/, scopes/, and lanes/ components before the date.
//! Each episode is one
//! immutable file per completed turn. Publication is: exclusive owner-only
//! temp file in the target directory → write → fsync → atomic no-replace
//! rename → parent directory fsync. An existing target is validated by
//! digest: exact duplicate is success, mismatch is a typed conflict.
//!
//! Containment: every component below the root is either a validated token
//! or generated here, and every descent refuses to follow symlinks, so a
//! link planted inside the corpus cannot redirect writes outside it.

const std = @import("std");
const Io = std.Io;
const contracts = @import("contracts.zig");
const identity = @import("identity.zig");
const render = @import("render.zig");

pub const dir_permissions: Io.File.Permissions = @enumFromInt(0o700);
pub const file_permissions: Io.File.Permissions = @enumFromInt(0o600);

pub const PublishError = error{
    /// A path component inside the corpus is a symlink or not a directory.
    ContainmentViolation,
    PermissionDenied,
    /// I/O failure or sync uncertainty; the caller may retry — idempotency
    /// makes redelivery safe.
    Unavailable,
    OutOfMemory,
};

pub const Published = struct {
    outcome: enum { published, duplicate, conflict },
    episode_id: [identity.episode_id_len]u8,
    digest_hex: [identity.digest_hex_len]u8,
    /// Owned by the caller's allocator: path relative to the journal root.
    rel_path: []u8,
    /// Owned by the caller's allocator: the rendered episode bytes, so the
    /// capture path can index without re-reading the file it just wrote.
    content: []u8,

    pub fn deinit(self: *const Published, gpa: std.mem.Allocator) void {
        gpa.free(self.rel_path);
        gpa.free(self.content);
    }
};

/// Publishes one validated payload into the journal root. The root `Dir` is
/// the configured journals root (already opened); the world subtree is
/// created on demand with owner-only permissions.
pub fn publish(
    root: Io.Dir,
    io: Io,
    gpa: std.mem.Allocator,
    payload: contracts.Payload,
    capture_time_ms: u64,
) PublishError!Published {
    const episode_id = identity.episodeId(payload);
    const digest_hex = identity.payloadDigestHex(payload);

    const content = render.render(gpa, .{
        .payload = payload,
        .episode_id = &episode_id,
        .digest_hex = &digest_hex,
        .capture_time_ms = capture_time_ms,
    }) catch return error.OutOfMemory;
    errdefer gpa.free(content);

    var shard: ShardPath = .init(payload.event_time_ms);
    var component_buf: [9][]const u8 = undefined;
    const components = layoutComponents(payload, &shard, &component_buf);
    var episode_dir = try openComponents(root, io, components);
    defer episode_dir.close(io);

    var final_name_buf: [identity.episode_id_len + 3]u8 = undefined;
    const final_name = std.fmt.bufPrint(&final_name_buf, "{s}.md", .{&episode_id}) catch unreachable;

    // `.` + id + `.` + up to 20 decimal ms digits + `.tmp` = id + 26; 48
    // leaves slack so the `catch unreachable` below stays trivially safe.
    var tmp_name_buf: [identity.episode_id_len + 56]u8 = undefined;
    var tmp_name: []const u8 = "";
    for (0..64) |attempt| {
        tmp_name = std.fmt.bufPrint(&tmp_name_buf, ".{s}.{d}.{d}.tmp", .{
            &episode_id, capture_time_ms, attempt,
        }) catch unreachable;
        writeTemp(episode_dir, io, tmp_name, content) catch |err| switch (err) {
            error.TempCollision => continue,
            error.ContainmentViolation => return error.ContainmentViolation,
            error.OutOfMemory => return error.OutOfMemory,
            error.PermissionDenied => return error.PermissionDenied,
            error.Unavailable => return error.Unavailable,
        };
        break;
    } else return error.Unavailable;
    var tmp_live = true;
    defer if (tmp_live) episode_dir.deleteFile(io, tmp_name) catch {};

    var outcome: @FieldType(Published, "outcome") = .published;
    episode_dir.renamePreserve(tmp_name, episode_dir, final_name, io) catch |err| switch (err) {
        error.PathAlreadyExists => {
            outcome = try classifyExisting(episode_dir, io, gpa, final_name, &digest_hex);
        },
        error.AccessDenied, error.PermissionDenied => return error.PermissionDenied,
        else => return error.Unavailable,
    };
    if (outcome == .published) tmp_live = false;

    // Make the directory entry durable before reporting success.
    syncDir(episode_dir, io) catch return error.Unavailable;

    var path_buf: [10][]const u8 = undefined;
    @memcpy(path_buf[0..components.len], components);
    path_buf[components.len] = final_name;
    const rel_path = std.mem.join(gpa, "/", path_buf[0 .. components.len + 1]) catch
        return error.OutOfMemory;

    return .{
        .outcome = outcome,
        .episode_id = episode_id,
        .digest_hex = digest_hex,
        .rel_path = rel_path,
        .content = content,
    };
}

/// `YYYY/MM/DD` derived from the source event time.
pub const ShardPath = struct {
    buf: [10]u8,

    pub fn init(event_time_ms: u64) ShardPath {
        const secs = std.time.epoch.EpochSeconds{ .secs = event_time_ms / 1000 };
        const year_day = secs.getEpochDay().calculateYearDay();
        const month_day = year_day.calculateMonthDay();
        var self: ShardPath = .{ .buf = undefined };
        _ = std.fmt.bufPrint(&self.buf, "{d:0>4}/{d:0>2}/{d:0>2}", .{
            @as(u16, year_day.year),
            @as(u8, month_day.month.numeric()),
            @as(u8, month_day.day_index) + 1,
        }) catch unreachable;
        return self;
    }

    pub fn slice(self: *const ShardPath) []const u8 {
        return &self.buf;
    }
};

fn layoutComponents(
    payload: contracts.Payload,
    shard: *const ShardPath,
    buf: *[9][]const u8,
) []const []const u8 {
    var len: usize = 0;
    if (!std.mem.eql(u8, payload.world, "main")) {
        buf[len] = "worlds";
        len += 1;
        buf[len] = payload.world;
        len += 1;
    }
    if (!std.mem.eql(u8, payload.scope, "default")) {
        buf[len] = "scopes";
        len += 1;
        buf[len] = payload.scope;
        len += 1;
    }
    if (payload.lane != .conversation) {
        buf[len] = "lanes";
        len += 1;
        buf[len] = @tagName(payload.lane);
        len += 1;
    }
    buf[len] = shard.buf[0..4];
    len += 1;
    buf[len] = shard.buf[5..7];
    len += 1;
    buf[len] = shard.buf[8..10];
    len += 1;
    return buf[0..len];
}

fn openComponents(root: Io.Dir, io: Io, components: []const []const u8) PublishError!Io.Dir {
    var current = root;
    var owns_current = false;
    errdefer if (owns_current) current.close(io);
    for (components) |component| {
        const next = try openOrCreateChild(current, io, component);
        if (owns_current) current.close(io);
        current = next;
        owns_current = true;
    }
    return current;
}

/// Opens a direct child directory without following symlinks, creating it
/// owner-only if absent. Concurrent creators are tolerated.
fn openOrCreateChild(parent: Io.Dir, io: Io, name: []const u8) PublishError!Io.Dir {
    const open_options: Io.Dir.OpenOptions = .{
        .follow_symlinks = false,
        .iterate = true,
    };
    var child = parent.openDir(io, name, open_options) catch |err| switch (err) {
        error.FileNotFound => blk: {
            parent.createDir(io, name, dir_permissions) catch |create_err| switch (create_err) {
                error.PathAlreadyExists => {},
                error.AccessDenied, error.PermissionDenied => return error.PermissionDenied,
                else => return error.Unavailable,
            };
            break :blk parent.openDir(io, name, open_options) catch |reopen_err|
                switch (reopen_err) {
                    error.NotDir, error.SymLinkLoop => return error.ContainmentViolation,
                    error.AccessDenied, error.PermissionDenied => return error.PermissionDenied,
                    else => return error.Unavailable,
                };
        },
        error.NotDir, error.SymLinkLoop => return error.ContainmentViolation,
        error.AccessDenied, error.PermissionDenied => return error.PermissionDenied,
        else => return error.Unavailable,
    };
    child.setPermissions(io, dir_permissions) catch |err| {
        child.close(io);
        return switch (err) {
            error.AccessDenied, error.PermissionDenied => error.PermissionDenied,
            else => error.Unavailable,
        };
    };
    return child;
}

const WriteTempError = PublishError || error{TempCollision};

fn writeTemp(dir: Io.Dir, io: Io, tmp_name: []const u8, content: []const u8) WriteTempError!void {
    var file = dir.createFile(io, tmp_name, .{
        .exclusive = true,
        .permissions = file_permissions,
        .resolve_beneath = true,
    }) catch |err| switch (err) {
        error.PathAlreadyExists => return error.TempCollision,
        error.AccessDenied, error.PermissionDenied => return error.PermissionDenied,
        else => return error.Unavailable,
    };
    // A failed write must remove the temp it created: the name embeds
    // `capture_time_ms`, so an orphan would make an identical redelivery
    // fail its exclusive create forever.
    errdefer {
        file.close(io);
        dir.deleteFile(io, tmp_name) catch {};
    }
    file.writeStreamingAll(io, content) catch return error.Unavailable;
    file.sync(io) catch return error.Unavailable;
    file.close(io);
}

fn classifyExisting(
    dir: Io.Dir,
    io: Io,
    gpa: std.mem.Allocator,
    final_name: []const u8,
    digest_hex: *const [identity.digest_hex_len]u8,
) PublishError!@FieldType(Published, "outcome") {
    var file = dir.openFile(io, final_name, .{}) catch |err| switch (err) {
        error.AccessDenied, error.PermissionDenied => return error.PermissionDenied,
        else => return error.Unavailable,
    };
    defer file.close(io);
    file.setPermissions(io, file_permissions) catch |err| switch (err) {
        error.AccessDenied, error.PermissionDenied => return error.PermissionDenied,
        else => return error.Unavailable,
    };
    const existing = dir.readFileAlloc(io, final_name, gpa, .limited(contracts.max_episode_file_bytes)) catch |err|
        switch (err) {
            error.OutOfMemory => return error.OutOfMemory,
            error.AccessDenied, error.PermissionDenied => return error.PermissionDenied,
            else => return error.Unavailable,
        };
    defer gpa.free(existing);
    const existing_digest = render.frontmatterDigestHex(existing) orelse return .conflict;
    return if (std.mem.eql(u8, existing_digest, digest_hex)) .duplicate else .conflict;
}

fn syncDir(dir: Io.Dir, io: Io) !void {
    var dir_file = try dir.openFile(io, ".", .{});
    defer dir_file.close(io);
    try dir_file.sync(io);
}

/// Counts authoritative-looking episode files under the journal root.
/// Diagnostics only; malformed candidates are excluded by sync.
pub fn countEpisodes(root: Io.Dir, io: Io) u64 {
    var corpus = root.openDir(io, ".", .{
        .follow_symlinks = false,
        .iterate = true,
    }) catch return 0;
    defer corpus.close(io);
    return countIn(corpus, io, contracts.corpus_walk_depth);
}

fn countIn(dir: Io.Dir, io: Io, depth_left: u8) u64 {
    var total: u64 = 0;
    var it = dir.iterate();
    while (it.next(io) catch return total) |entry| {
        switch (entry.kind) {
            .file => if (std.mem.startsWith(u8, entry.name, "aj1-") and
                std.mem.endsWith(u8, entry.name, ".md"))
            {
                total += 1;
            },
            .directory => if (depth_left > 0) {
                // Same visibility rule as the index walk: dot-directories
                // are foreign tooling state, never episode shards.
                if (entry.name.len == 0 or entry.name[0] == '.') continue;
                var child = dir.openDir(io, entry.name, .{
                    .follow_symlinks = false,
                    .iterate = true,
                }) catch continue;
                defer child.close(io);
                total += countIn(child, io, depth_left - 1);
            },
            else => {},
        }
    }
    return total;
}

File: adapters/pi/engine/store.ts
// Atomic episode publication under a contained journal root, and the
// whole capture transaction.
//
// Default layout: YYYY/MM/DD/<episode-id>.md. Non-default classifications
// add reserved worlds/, scopes/, and lanes/ components before the date.
// Each episode is one immutable file per completed turn. Publication is:
// exclusive owner-only temp file in the target directory → write → fsync →
// atomic no-replace hard-link into place → temp unlink → parent directory
// fsync. An existing target is classified by recorded digest: exact
// duplicate is success, anything else is a typed conflict. (v1's
// supersede path — in-place replacement of a proven extension — was
// removed in v2: capture fires once per settled turn, so a same-identity
// redelivery with different bytes only ever means divergence.)
//
// The descent, temp-write, and fsync mechanics live in corpus.ts with the
// rest of the containment discipline; this file owns the publication
// decision itself.

import * as fs from "node:fs";
import * as path from "node:path";
import {
  MAX_CONTENT_BYTES,
  MAX_EPISODE_FILE_BYTES,
  validate,
  CaptureError,
  captureErrorName,
  type CaptureOutcome,
  type IndexFreshness,
  type Payload,
  type RawPayload,
  type CaptureErrorCode,
} from "./contracts.ts";
import { episodeId, payloadDigestHex } from "./identity.ts";
import { render } from "./render.ts";
import { frontmatterDigestHex } from "./render.ts";
import type { CaptureDefaults } from "./config.ts";
import {
  descendCreating,
  layoutComponents,
  openJournalRoot,
  rootInSharedDirectory,
  syncDir,
  writeTemp,
  StoreError,
  TempCollisionError,
  type JournalRoot,
} from "./corpus.ts";
import { resolveJournalRoot } from "./paths.ts";

/** The result of one publish call. */
export interface Published {
  outcome: Extract<CaptureOutcome, "published" | "duplicate" | "conflict">;
  episodeId: string;
  digestHex: string;
  /**
   * The episode path relative to the journal root, slash-joined (the path
   * vocabulary of evidence references).
   */
  relPath: string;
  /**
   * The rendered episode content, so the capture path can index without
   * re-reading the file it just wrote.
   */
  content: string;
}

/** Oversize-policy accounting carried alongside a publish. */
export interface DroppedBytes {
  user: number;
  assistant: number;
}

const NO_DROPS: DroppedBytes = { user: 0, assistant: 0 };

/**
 * Publishes one validated payload into the journal root. The world
 * subtree is created on demand with owner-only permissions.
 */
export function publish(
  root: JournalRoot,
  payload: Payload,
  captureTimeMs: number,
  drops: DroppedBytes = NO_DROPS,
): Published {
  const id = episodeId(payload);
  const digestHex = payloadDigestHex(payload);
  const content = render({
    payload,
    episodeId: id,
    digestHex,
    captureTimeMs,
    userDroppedBytes: drops.user,
    assistantDroppedBytes: drops.assistant,
  });
  const contentBytes = Buffer.from(content, "utf8");

  const components = layoutComponents(payload);
  const episodeDir = descendCreating(root, components);

  const finalName = id + ".md";
  // The temp name embeds the capture time and an attempt counter; a
  // collision (orphan from a crashed writer) retries a fresh name.
  let tmpName = "";
  let written = false;
  for (let attempt = 0; attempt < 64 && !written; attempt++) {
    tmpName = `.${id}.${captureTimeMs}.${attempt}.tmp`;
    try {
      writeTemp(episodeDir, tmpName, contentBytes);
      written = true;
    } catch (err) {
      if (err instanceof TempCollisionError) continue;
      throw err;
    }
  }
  if (!written) throw new StoreError("unavailable", "temp name: 64 collisions");

  let outcome: Published["outcome"] = "published";
  try {
    try {
      fs.linkSync(path.join(episodeDir, tmpName), path.join(episodeDir, finalName));
    } catch (err) {
      if ((err as NodeJS.ErrnoException).code !== "EEXIST") {
        throw classifyIo("link episode", err);
      }
      outcome = classifyExisting(episodeDir, finalName, digestHex);
    }
  } finally {
    // Whatever happened above, the temp file does not outlive this call.
    try {
      fs.rmSync(path.join(episodeDir, tmpName), { force: true });
    } catch {
      // Best-effort: an orphan is invisible to readers.
    }
  }

  // Make the directory entry durable before reporting success.
  syncDir(episodeDir);

  return {
    outcome,
    episodeId: id,
    digestHex,
    relPath: [...components, finalName].join("/"),
    content,
  };
}

function classifyIo(context: string, err: unknown): StoreError {
  const code = (err as NodeJS.ErrnoException).code;
  if (code === "EACCES" || code === "EPERM") return new StoreError("permission_denied", `${context}: ${String(err)}`);
  return new StoreError("unavailable", `${context}: ${String(err)}`);
}

// classifyExisting decides duplicate or conflict for a target that already
// exists at the derived path: the stored file's *recorded* digest equal to
// the incoming digest is a duplicate — deliberately the recorded line, not
// a re-verification, so an exact redelivery of an episode the owner has
// since hand-edited stays a duplicate, which is the answer that keeps
// redelivery idempotent. Anything else — a differing digest, a file with
// no readable digest line — is a conflict.
//
// The existing file's permissions are repaired to owner-only on the way:
// owner-only is a standing invariant, and a redelivery is a free
// opportunity to fix a file that lost it.
function classifyExisting(dirAbs: string, finalName: string, digestHex: string): "duplicate" | "conflict" {
  const finalPath = path.join(dirAbs, finalName);
  let existing: Buffer;
  try {
    fs.chmodSync(finalPath, 0o600);
    existing = fs.readFileSync(finalPath);
  } catch (err) {
    throw classifyIo("read existing episode", err);
  }
  if (existing.byteLength > MAX_EPISODE_FILE_BYTES) {
    throw new StoreError("unavailable", `read existing episode: exceeds ${MAX_EPISODE_FILE_BYTES} bytes`);
  }
  const recorded = frontmatterDigestHex(existing.toString("utf8"));
  if (recorded !== null && recorded === digestHex) return "duplicate";
  return "conflict";
}

/**
 * The oversize policy (owner ruling 2026-08-31): a side over the content
 * budget is deterministically tail-truncated to the largest code-point
 * boundary within the budget instead of rejecting the turn, and the
 * dropped byte count is recorded in frontmatter rather than vanishing.
 * Returns the payload to validate plus the per-side accounting.
 */
export function applyOversizePolicy(raw: RawPayload): { raw: RawPayload; drops: DroppedBytes } {
  const [userContent, user] = truncateTail(raw.userContent);
  const [assistantResult, assistant] = truncateTail(raw.assistantResult);
  if (user === 0 && assistant === 0) return { raw, drops: NO_DROPS };
  return { raw: { ...raw, userContent, assistantResult }, drops: { user, assistant } };
}

// truncateTail cuts a string to MAX_CONTENT_BYTES of UTF-8, backing off to
// a code-point boundary, and reports the dropped byte count. Lone
// surrogates would make the byte accounting ill-defined; they are left for
// validate to refuse as InvalidUtf8.
function truncateTail(s: string): [string, number] {
  const total = Buffer.byteLength(s, "utf8");
  if (total <= MAX_CONTENT_BYTES || !s.isWellFormed()) return [s, 0];
  const bytes = Buffer.from(s, "utf8");
  let cut = MAX_CONTENT_BYTES;
  // Back off past UTF-8 continuation bytes (10xxxxxx) to a boundary.
  while (cut > 0 && (bytes[cut] & 0b1100_0000) === 0b1000_0000) cut--;
  const kept = bytes.subarray(0, cut).toString("utf8");
  return [kept, total - cut];
}

/** One whole capture transaction's input. */
export interface CaptureInput {
  rootPath: string;
  raw: RawPayload;
  /** Owner capture defaults, for world/scope fill. */
  defaults: CaptureDefaults;
  captureTimeMs: number;
}

/**
 * The transaction's typed outcome. detail carries the capture error code
 * for failure outcomes and is empty for every success.
 */
export interface CaptureResult {
  outcome: CaptureOutcome;
  episodeId: string;
  digestHex: string;
  relPath: string;
  indexState: IndexFreshness;
  detail: CaptureErrorCode | "";
  /**
   * True when the refusal was the shared-directory rule specifically, so
   * a renderer can keep its long-standing remediation wording. Wording is
   * rendering; this flag is the typed sentinel.
   */
  sharedDirectory: boolean;
}

/**
 * Composes the whole capture transaction so the extension and the CLI run
 * the same code rather than the same intent: defaults fill, the oversize
 * policy, validate, root canonicalization, shared-directory refusal,
 * atomic publication. The order is part of the contract: shared-directory
 * refusal is decided before the root is opened (a refused root is never
 * created), and validate before either.
 *
 * The index projection joins this composition at the snapshot-index slice;
 * until then successful publication honestly reports a stale projection.
 */
export function capture(input: CaptureInput): CaptureResult {
  const failure = (outcome: CaptureOutcome, detail: CaptureErrorCode, sharedDirectory = false): CaptureResult => ({
    outcome,
    episodeId: "",
    digestHex: "",
    relPath: "",
    indexState: "not_built",
    detail,
    sharedDirectory,
  });

  // Owner-default world/scope fill: a host provides explicit values only
  // when transporting an owner session choice.
  let raw = input.raw;
  if (raw.world === null) raw = { ...raw, world: input.defaults.world };
  if (raw.scope === null) raw = { ...raw, scope: input.defaults.scope };
  const sized = applyOversizePolicy(raw);

  let payload: Payload;
  try {
    payload = validate(sized.raw);
  } catch (err) {
    return failure("malformed", captureErrorName(err));
  }

  const rootPath = resolveJournalRoot(input.rootPath);
  if (rootInSharedDirectory(rootPath)) {
    return failure("permission_denied", "PermissionDenied", true);
  }

  let root: JournalRoot;
  try {
    root = openJournalRoot(rootPath);
  } catch (err) {
    return failure("unavailable", storeErrorCode(err));
  }

  let published: Published;
  try {
    published = publish(root, payload, input.captureTimeMs, sized.drops);
  } catch (err) {
    if (err instanceof StoreError) {
      switch (err.code) {
        case "containment_violation":
          return failure("internal_error", "ContainmentViolation");
        case "permission_denied":
          return failure("permission_denied", "PermissionDenied");
        default:
          return failure("unavailable", "Unavailable");
      }
    }
    return failure("unavailable", "Unavailable");
  }

  return {
    outcome: published.outcome,
    episodeId: published.episodeId,
    digestHex: published.digestHex,
    relPath: published.relPath,
    indexState: "stale",
    detail: "",
    sharedDirectory: false,
  };
}

function storeErrorCode(err: unknown): CaptureErrorCode {
  if (err instanceof StoreError && err.code === "permission_denied") return "PermissionDenied";
  if (err instanceof CaptureError) return err.code;
  return "Unavailable";
}

